pkg/router/controller/router_controller.go

```diff
@@ -183,56 +105,36 @@ func (c *RouterController) HandleNode() {
 	if err := c.Plugin.HandleNode(eventType, node); err != nil {
 		utilruntime.HandleError(err)
 	}
+	c.updateConsumedCount(node)
 }
 
 // HandleRoute handles a single Route event and synchronizes the router backend.
-func (c *RouterController) HandleRoute() {
-	eventType, route, err := c.NextRoute()
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("unable to read routes: %v", err))
-		return
-	}
-
+func (c *RouterController) HandleRoute(eventType watch.EventType, obj interface{}) {
+	route := obj.(*routeapi.Route)
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
 	c.processRoute(eventType, route)
-
-	// Change the local sync state within the lock to ensure that all
-	// event handlers have the same view of sync state.
-	c.routesListConsumed = c.RoutesListConsumed()
-	c.commit()
+	c.updateConsumedCount(route)
+	c.Commit()
 }
 
 // HandleEndpoints handles a single Endpoints event and refreshes the router backend.
-func (c *RouterController) HandleEndpoints() {
-	eventType, endpoints, err := c.NextEndpoints()
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("unable to read endpoints: %v", err))
-		return
-	}
-
+func (c *RouterController) HandleEndpoints(eventType watch.EventType, obj interface{}) {
+	endpoints := obj.(*kapi.Endpoints)
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
 	if err := c.Plugin.HandleEndpoints(eventType, endpoints); err != nil {
 		utilruntime.HandleError(err)
 	}
-
-	// Change the local sync state within the lock to ensure that all
-	// event handlers have the same view of sync state.
-	c.endpointsListConsumed = c.EndpointsListConsumed()
-	c.commit()
+	c.updateConsumedCount(endpoints)
```

Comment:
Counting isn't enough to tell you whether you've got everything.  If an endpoint is deleted while you're doing the sync you will never reach your target number.  You can't synchronize with your handlers like this unfortunately.